db/memtable.cc

```diff
@@ -1011,4 +1012,21 @@ uint64_t MemTable::GetMinLogContainingPrepSection() {
   return min_prep_log_referenced_.load();
 }
 
+bool MemTable::ShouldFlushAfterSampledRead() {
+  // TODO: Make the two constants options.
+  static const int kSampleRate = 1024;
+  static const int kNumReadsUntilFlush = 1024 * 1024;
+
+  if (!Random::GetTLSInstance()->OneIn(kSampleRate)) {
+    return false;
+  }
+  uint64_t num_reads = num_reads_sampled_.fetch_add(kSampleRate) + kSampleRate;
+  do {
+    if (num_reads < kNumReadsUntilFlush) {
+      return false;
+    }
+  } while (!num_reads_sampled_.compare_exchange_weak(num_reads, 0LL));
```

Comment:
`num_reads` may keep changing what's so ever, right? I know always do a CAS is more expensive than a counter increment, but it feels that doing both of atomic incremental and CAS over-complicates the logic, and make it harder to reason about correctness.
Did you consider to only do CAS?

Reply:
It's not immediately clear to me how to do this in a simpler way with only CAS. The increment handles the common case when we sample, the CAS handles the case where the counter is reset to 0 (only once).

Reply:
Just do CAS for common case too. CAS incremental and CAS reset to 0.

Reply:
Increment via CAS seems more complicated than regular increment?

Reply:
Something like this:
```
uint64_t num_reads;
uint64_t new_num_reads;
bool ret;
do {
  num_reads = num_reads_sampled_.load();
  new_num_reads = num_reads + kSampleRate;
  if (new_num_reads  < kNumReadsUntilFlush) {
      ret = false;
   } else {
     new_num_reads  = 0;
     ret  = true;
   }
} while (!num_reads_sampled_.compare_exchange_weak(num_reads, new_num_reads));
return ret;
```